modula/torch_modules.py

```diff
@@ -0,0 +1,330 @@
+
+
+def orthogonalize(M):
+    # six step Newton-Schulz by @YouJiacheng
+    # coefficients from: https://twitter.com/YouJiacheng/status/1893704552689303901
+    # found by optimization: https://gist.github.com/YouJiacheng/393c90cbdc23b09d5688815ba382288b/5bff1f7781cf7d062a155eecd2f13075756482ae
+    # the idea of stability loss was from @leloykun
+
+    abc_list = [
+        (3955/1024, -8306/1024, 5008/1024),
+        (3735/1024, -6681/1024, 3463/1024),
+        (3799/1024, -6499/1024, 3211/1024),
+        (4019/1024, -6385/1024, 2906/1024),
+        (2677/1024, -3029/1024, 1162/1024),
+        (2172/1024, -1833/1024,  682/1024)
+    ]
+
+    transpose = M.shape[1] > M.shape[0]
+    if transpose:
+        M = M.T
+    M = M / torch.linalg.norm(M)
+    for a, b, c in abc_list:
+        A = M.T @ M
+        I = torch.eye(A.shape[0], device=M.device, dtype=M.dtype)
+        M = M @ (a * I + b * A + c * A @ A)
+    if transpose:
+        M = M.T
+    return M
+
+
+# Atomic
+class Linear(nn.Linear):
+    def __init__(self, fanin, fanout, target_norm=1.0):
+        super().__init__(fanin, fanout, bias=False)
+        self.fanin = fanin
+        self.fanout = fanout
+        self.register_buffer('target_norm', torch.tensor(target_norm, dtype=torch.float))
+        
+        # Initialize with orthogonal weights
+        self._initialize_weights()
+    
+    def _initialize_weights(self):
+        with torch.no_grad():
+            weight = torch.randn(self.fanout, self.fanin)
+            weight = orthogonalize(weight) * torch.sqrt(torch.tensor(self.fanout / self.fanin))
+            self.weight.copy_(weight)
+    
+    def project_weights(self):
+        """Project weights to the constraint manifold"""
+        with torch.no_grad():
+            weight = orthogonalize(self.weight) * torch.sqrt(torch.tensor(self.fanout / self.fanin))
+            self.weight.copy_(weight)
+    
+    def dualize_gradients(self):
+        """Apply dualization to gradients"""
+        if self.weight.grad is not None:
+            with torch.no_grad():
+                grad = self.weight.grad
+                d_weight = orthogonalize(grad) * torch.sqrt(torch.tensor(self.fanout / self.fanin)) * self.target_norm
+                self.weight.grad.copy_(d_weight)
+
+    @staticmethod
+    def from_modula(m, w=None):
+        """Convert from modula.atom.Linear"""
+        with torch.no_grad():
+            linear = Linear(m.fanin, m.fanout, getattr(m, 'target_norm', 1.0))
+            if w is not None:
+                linear.weight.copy_(w)
+        return linear
+
+    def __repr__(self):
+        return f"Linear(fanin={self.fanin}, fanout={self.fanout}, target_norm={self.target_norm})"
+
+class Embed(nn.Embedding):
+    def __init__(self, num_embed, d_embed, target_norm, padding_idx=None):
+        super().__init__(num_embed, d_embed, padding_idx=padding_idx)
+        self.num_embed = num_embed
+        self.d_embed = d_embed
+        self.sensitivity = 1
```

Comment:
I was trying not to include `sensitivity` attributes that don't do anything (the target norms are inherited from the Jax module and don't do anything after conversion).